opal/mca/smsc/accelerator/smsc_accelerator_module.c

```diff
@@ -0,0 +1,227 @@
+                                     void *remote_ptr, size_t size, void **local_ptr)
+{
+    void* result = NULL;
+    mca_smsc_accelerator_endpoint_t *ep = (mca_smsc_accelerator_endpoint_t *)endpoint;
+
+    if (NULL != mca_smsc_accelerator_module.prev_smsc &&
+        (mca_smsc_accelerator_module.prev_smsc->features & MCA_SMSC_FEATURE_CAN_MAP)) {
+        result = mca_smsc_accelerator_module.prev_smsc->map_peer_region(ep->prev_endpoint,
+                                                                        flags, remote_ptr, size, local_ptr);
+    }
+    return result;
+}
+
+void mca_smsc_accelerator_unmap_peer_region(void *ctx)
+{
+    if (NULL != mca_smsc_accelerator_module.prev_smsc &&
+        (mca_smsc_accelerator_module.prev_smsc->features & MCA_SMSC_FEATURE_CAN_MAP)) {
+        mca_smsc_accelerator_module.prev_smsc->unmap_peer_region(ctx);
+    }
+}
+
+static int mca_smsc_accelerator_copy (mca_smsc_endpoint_t *endpoint, void *local_address, void *remote_address,
+                                      size_t size, void *reg_handle, bool is_remote_write)
+{
+    mca_smsc_accelerator_endpoint_t *ep = (mca_smsc_accelerator_endpoint_t *)endpoint;
+    mca_smsc_accelerator_registration_data_t *reg = (mca_smsc_accelerator_registration_data_t *)reg_handle;
+    int ret = OPAL_SUCCESS;
+    void *remote_memory_address;
+    mca_opal_gpu_reg_t rget_reg;
+    mca_opal_gpu_reg_t *reg_ptr = &rget_reg;
+    size_t offset;
+
+    memset(&rget_reg, 0, sizeof(rget_reg));
+    memcpy(&rget_reg.data.ipcHandle.handle, reg->handle.accelerator,
+           SMSC_ACCELERATOR_HANDLE_SIZE);
+    ret = ep->rcache->rcache_register(ep->rcache, remote_address, size,
+                                      /*ep->peer_smp_rank for debugging only */ 0,
+                                      MCA_RCACHE_ACCESS_LOCAL_WRITE,
+                                      (mca_rcache_base_registration_t **) &reg_ptr);
+    if (OPAL_SUCCESS != ret) {
+        opal_output(0, "mca_smsc_accelerator_copy: Failed to register remote memory, ret=%d", ret);
+        return ret;
+    }
+
+    offset = (size_t)((intptr_t) remote_address - (intptr_t) reg->base_addr);
+    remote_memory_address = (unsigned char *) reg_ptr->base.alloc_base + offset;
+
+    if (is_remote_write) {
+        ret = opal_accelerator.mem_copy(MCA_ACCELERATOR_NO_DEVICE_ID, mca_smsc_accelerator_module.device_id,
+                                        remote_memory_address, local_address, size,
+                                        MCA_ACCELERATOR_TRANSFER_DTOD);
+    } else {
+        ret = opal_accelerator.mem_copy(mca_smsc_accelerator_module.device_id, MCA_ACCELERATOR_NO_DEVICE_ID,
+                                        local_address, remote_memory_address, size,
+                                        MCA_ACCELERATOR_TRANSFER_DTOD);
+    }
+    if (OPAL_SUCCESS != ret) {
+        opal_output(0, "mca_smsc_accelerator_copy: accelerator mem_copy failed, ret=%d", ret);
```

Comment:
It would be more interesting if we could report more information about the error here, maybe move this in each branch of the if and print the important `mem_copy` arguments.

Reply:
@bosilca thank you for your review, I incorporated your suggestion!